opensearch-operator/pkg/reconcilers/ismpolicy.go

```diff
@@ -0,0 +1,575 @@
+package reconcilers
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"github.com/go-logr/logr"
+	"github.com/opensearch-project/opensearch-go/opensearchutil"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/tools/record"
+	"k8s.io/client-go/util/retry"
+	"k8s.io/utils/pointer"
+	opsterv1 "opensearch.opster.io/api/v1"
+	"opensearch.opster.io/opensearch-gateway/requests"
+	"opensearch.opster.io/opensearch-gateway/services"
+	"opensearch.opster.io/pkg/reconcilers/util"
+	"reflect"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+	"time"
+)
+
+const (
+	ismPolicyExists = "ism policy already exists in Opensearch"
+	ismResource     = "_ism"
+)
+
+type IsmPolicyReconciler struct {
+	client.Client
+	ReconcilerOptions
+	ctx      context.Context
+	osClient *services.OsClusterClient
+	recorder record.EventRecorder
+	instance *opsterv1.ISMPolicy
+	cluster  *opsterv1.OpenSearchCluster
+	logger   logr.Logger
+}
+
+func NewIsmReconciler(
+	ctx context.Context,
+	client client.Client,
+	recorder record.EventRecorder,
+	instance *opsterv1.ISMPolicy,
+	opts ...ReconcilerOption,
+) *IsmPolicyReconciler {
+	options := ReconcilerOptions{}
+	options.apply(opts...)
+	return &IsmPolicyReconciler{
+		Client:            client,
+		ReconcilerOptions: options,
+		ctx:               ctx,
+		recorder:          recorder,
+		instance:          instance,
+		logger:            log.FromContext(ctx).WithValues("ismpolicy", "ism"),
+	}
+}
+
+func (r *IsmPolicyReconciler) Reconcile() (retResult ctrl.Result, retErr error) {
+	var reason string
+	defer func() {
+		if !pointer.BoolDeref(r.updateStatus, true) {
+			return
+		}
+		// When the reconciler is done, figure out what the state of the resource
+		// is and set it in the state field accordingly.
+		err := retry.RetryOnConflict(retry.DefaultRetry, func() error {
+			if err := r.Get(r.ctx, client.ObjectKeyFromObject(r.instance), r.instance); err != nil {
+				return err
+			}
+			r.instance.Status.Reason = reason
+			if retErr != nil {
+				r.instance.Status.State = opsterv1.OpensearchISMPolicyError
+			}
+			// Requeue after is 10 seconds if waiting for OpenSearch cluster
+			if retResult.Requeue && retResult.RequeueAfter == 10*time.Second {
+				r.instance.Status.State = opsterv1.OpensearchISMPolicyPending
+			}
+			// Requeue is after 30 seconds for normal reconciliation after creation/update
+			if retErr == nil && retResult.RequeueAfter == 30*time.Second {
+				r.instance.Status.State = opsterv1.OpensearchISMPolicyCreated
+			}
+			if reason == ismPolicyExists {
+				r.instance.Status.State = opsterv1.OpensearchISMPolicyIgnored
+			}
+			return r.Status().Update(r.ctx, r.instance)
+		})
+
+		if err != nil {
+			r.logger.Error(err, "failed to update status")
+		}
+	}()
+
```

Comment:
There should be a check here, like in the other reconcilers, to verify the cluster reference has not changed.